src/data/finsum.lean

```diff
@@ -0,0 +1,727 @@
+  ∑ᶠ i in s, f i + ∑ᶠ i in t \ s, f i = ∑ᶠ i in t, f i :=
+begin
+  conv_rhs { rw ← set.union_diff_cancel hst },
+  exact (finsum_in_union (set.finite.subset ht hst)
+    (set.finite.subset ht $ set.diff_subset t s) set.disjoint_diff).symm
+end
+
+/-- A more general version of `finsum_in_sdiff` that requires `t ∩ function.support f` instead of
+  `t` to be finite. -/
+lemma finsum_in_sdiff' (ht : (t ∩ function.support f).finite)
+  (hst : s ⊆ t) : ∑ᶠ i in s, f i + ∑ᶠ i in t \ s, f i = ∑ᶠ i in t, f i :=
+begin
+  conv_rhs { rw ← set.union_diff_cancel hst },
+  exact (finsum_in_union' (set.finite.subset ht (set.inter_subset_inter_left _ hst))
+    (set.finite.subset ht (set.inter_subset_inter (set.diff_subset t s) $ set.subset.refl _))
+      set.disjoint_diff).symm
+end
+
+/-- Given the function `t : β → set α` in which for all `b : β`, `t b` is finite and for all
+  distinct `x y ∈ s`, `t x` and `t y` are disjoint, then the sum on `⋃ x ∈ s, t x` over the
+  function `f` equals the sum on `s` over `λ x, finsum_in (t x) f`. -/
+lemma finsum_in_bUnion {s : set β} {t : β → set α} (hs : s.finite)
+  (ht : ∀ b, (t b).finite) (h : ∀ x ∈ s, ∀ y ∈ s, x ≠ y → disjoint (t x) (t y)) :
+  ∑ᶠ i in (⋃ x ∈ s, t x), f i = ∑ᶠ i in s, (∑ᶠ j in (t i), f j) :=
+begin
+  rw finsum_in_eq_finite_to_finset_sum _ hs,
+  rw finsum_in_eq_finite_to_finset_sum f (set.finite.bUnion hs (λ i _, ht i)),
+  conv_rhs { congr, skip, funext, rw finsum_in_eq_finite_to_finset_sum f (ht i) },
+  convert @finset.sum_bUnion _ _ _ f _ _ hs.to_finset (λ x, (ht x).to_finset)
+    (begin
+      intros x hx y hy hxy a ha,
+      specialize h x ((set.finite.mem_to_finset _).1 hx) y ((set.finite.mem_to_finset _).1 hy) hxy,
+      apply @h a, simpa using ha
+    end),
+  ext, rw [finset.mem_bUnion, set.finite.mem_to_finset, set.mem_bUnion_iff],
+  split; intro ha; rcases ha with ⟨x, hx₀, hx₁⟩,
+    exact ⟨x, (set.finite.mem_to_finset _).2 hx₀, (set.finite.mem_to_finset _).2 hx₁⟩,
+    exact ⟨x, (set.finite.mem_to_finset _).1 hx₀, (set.finite.mem_to_finset _).1 hx₁⟩
+end .
+
+/-- An alternative version of `finsum_in_bUnion` in which we sum on the a fintype rather than a
+  finite set. -/
+lemma finsum_in_Union [fintype β] {t : β → set α}
+  (ht : ∀ b, (t b).finite) (h : ∀ x y, x ≠ y → disjoint (t x) (t y)) :
+  ∑ᶠ i in (⋃ x : β, t x), f i = ∑ᶠ i, (∑ᶠ j in (t i), f j) :=
+begin
+  rw ← finsum_eq_finsum_in_univ,
+  rw ← finsum_in_bUnion (set.finite.of_fintype _) ht (λ x _ y _, h x y),
+    { congr, funext y, simp only [set.Union_pos, set.mem_univ] },
+end
+
+/-- An alternative version of `finsum_in_bUnion` in which `t` is a finite set of `set α`s. -/
+lemma finsum_in_sUnion {t : set (set α)} (ht₀ : t.finite)
+  (ht₁ : ∀ (x : set α), x.finite) (h : ∀ x ∈ t, ∀ y ∈ t, x ≠ y → disjoint x y) :
+  ∑ᶠ i in (⋃₀ t), f i = ∑ᶠ s in t, (∑ᶠ i in s, f i) :=
+by rw [set.sUnion_eq_bUnion, finsum_in_bUnion ht₀ ht₁ (by simpa)];
+  exact finsum_in_congr rfl (λ x hx, dif_pos hx)
+
```

Comment:
I don't believe that `ht₁` just above is what you want, since this just amounts to `(fintype \a)`. You really want the weaker hypothesis that each set in t is finite. This probably makes the proof nastier - here is a not-very-golfy proof I wrote with that adjusted hypothesis.

```
lemma finsum_in_sUnion' {t : set (set α)} (ht₀ : t.finite)
  (ht₁ : ∀ x ∈ t, set.finite x) (h : set.pairwise_disjoint t):
  ∑ᶠ i in (⋃₀ t), f i = ∑ᶠ s in t, (∑ᶠ i in s, f i) :=
begin
  rw [set.sUnion_eq_bUnion], 
  have h' := @finsum_in_bUnion α M _ t f set.univ subtype.val 
    (set.univ_finite_iff_nonempty_fintype.mpr ⟨ht₀.fintype⟩) 
    (λ s, ht₁ s.1 s.2) 
    (λ x hx y hy hxy, h x.1 x.2 y.1 y.2 (λ hn, hxy (subtype.val_injective hn))),
  rw [← finsum_eq_finsum_in_univ] at h', 
  nth_rewrite 1 ← finsum_subtype_eq_finsum_in, 
  erw ← h', 
  congr' 1,
  ext, simp,
end 
```

I also changed hypothesis `h` to use `set.pairwise_disjoint`.

Reply:
Thanks! You're surely right. Furthermore the branch isn't compiling for some reason right now. I'll fix up these things soon.

Reply:
I have replaced the lemma with your version.